File: Coding-Patterns/Coding-Patterns-CSharp/CodingPatterns/05_greedy/MergeIntervals.cs
// LeetCode 56 - Merge Intervals
// Difficulty: Medium
// Pattern: Sort + linear sweep
//
// Time: O(n log n)  Space: O(n)

namespace CodingPatterns.Greedy;

public class MergeIntervals {
    public int[][] Merge(int[][] intervals) {
        Array.Sort(intervals, (a, b) => a[0] - b[0]);
        LinkedList<int[]> merged = new LinkedList<int[]>();
        foreach (int[] interval in intervals) {
            // if the list of merged intervals is empty or if the current
            // interval does not overlap with the previous, append it
            if (merged.Count == 0 || merged.Last.Value[1] < interval[0]) {
                merged.AddLast(interval);
            }
            // otherwise, there is overlap, so we merge the current and previous
            // intervals
            else {
                merged.Last.Value[1] =
                    Math.Max(merged.Last.Value[1], interval[1]);
            }
        }

        return merged.ToArray();
    }

    public static void Run() {
        var sol = new MergeIntervals();
        int[][] intervals = new int[][] {
            new int[] {1, 3},
            new int[] {2, 6},
            new int[] {8, 10},
            new int[] {15, 18}
        };
        var merged = sol.Merge(intervals);
        Console.WriteLine("Merged intervals:");
        foreach (var interval in merged) {
            Console.WriteLine($"[{interval[0]}, {interval[1]}]");
        }
    }
}

File: Coding-Patterns/Coding-Patterns-Python/05_greedy/merge_intervals.py
# LeetCode 56 - Merge Intervals
# Difficulty: Medium
# Pattern: Sort + linear sweep
#
# Problem: Given an array of intervals, merge all overlapping intervals.
#
# Approach: Sort by start time. Walk through intervals; if current start
# <= last merged end, extend the end. Otherwise, add new interval.
#
# Time: O(n log n)  Space: O(n)

from typing import List

def merge(intervals: List[List[int]]) -> List[List[int]]:
    intervals.sort(key=lambda x: x[0])
    merged = [intervals[0]]

    for start, end in intervals[1:]:
        if start <= merged[-1][1]:
            merged[-1][1] = max(merged[-1][1], end)
        else:
            merged.append([start, end])

    return merged


# ---- Tests ----
if __name__ == "__main__":
    assert merge([[1,3],[2,6],[8,10],[15,18]]) == [[1,6],[8,10],[15,18]]
    assert merge([[1,4],[4,5]]) == [[1,5]]
    assert merge([[1,4],[2,3]]) == [[1,4]]
    print("All tests passed.")
